### PR TITLE
Kuangyi

### PR DESCRIPTION
**Pull request by KuangyiXing**

### CODE REVIEW

**Comment by KuangyiXing** [comment]
Engine.cs is not written by me.

File: Snake/BusinessLayer/Engine.cs [hunk 1]
```diff
@@ -1,217 +1,217 @@
...
-                return false;
-            }
-            if (mazeArray[x, y] == (int)Elements.snakeHead)
-            {
-                return false;
-            }
-            return true;
-        }
-    }
+﻿using BusinessLayer.FoodFolder;
```
Comment (danaterr) [inline Snake/BusinessLayer/Engine.cs]: change your sort usings settings in visual studio
convention is system first in alphabetical then the rest is alphabetical. there is a setting that will do this in VS

File: Snake/BusinessLayer/Engine.cs [hunk 2]
```diff
@@ -1,217 +1,217 @@
...
+        // For future use, 2 player game mode
+        //private GameSnake gameSnake2;
+        private Food food;
+        private FoodGenerator foodGenerator;
+
+        private enum gameMode
+        {
+            basic,
+        }
+        public enum snakeAction
```
Comment (danaterr) [inline Snake/BusinessLayer/Engine.cs]: public enums should be in their own files

File: Snake/BusinessLayer/Engine.cs [hunk 3]
```diff
@@ -1,217 +1,217 @@
...
+                    do
+                    {
+                        food = foodGenerator.generateFood(mazeLength, mazeWidth);
+                        isValid = validateNewFoodLocation(food);
+                    } while (!isValid);
+                    mazeArray[food.xLocation, food.yLocation] = Food;
+                    break;
+
+                default:
+                    throw new System.Exception("Invalid Game Mode!");
```
Comment (danaterr) [inline Snake/BusinessLayer/Engine.cs]: only fully qualify a class with the namespace to avoid a collision

File: Snake/BusinessLayer/Engine.cs [hunk 4]
```diff
@@ -1,217 +1,217 @@
...
+                    if (Score.getScore() > Score.getHighScore())
+                    {
+                        Score.setHighScore(Score.getScore());
+                    }
+                    mazeArray[0, 0] = snakeHitsMaze;
+                    return mazeArray;
+
+
+                case Food:  // snake hits the food
+                    snakesNewLocation = gameSnake1.snakeMove(snakeDirection, true);
```
Comment (danaterr) [inline Snake/BusinessLayer/Engine.cs]: this is a big chunk of code of a case. consider if some of this should be in their own methods with single responsibilities

File: Snake/BusinessLayer/Engine.cs [hunk 5]
```diff
@@ -1,217 +1,217 @@
...
+
+                    if ((newSnakeHead.returnX() == food.xLocation) && (newSnakeHead.returnY() == food.yLocation))
+                    {
+                        Score.incrementScore(food.pointsWorth);
+
+                    }
+
+                    food = null;
+                    bool isValid = true;
+                    do
```
Comment (danaterr) [inline Snake/BusinessLayer/Engine.cs]: i have seen this code else where. should it be in its own method

File: Snake/BusinessLayer/Engine.cs [hunk 6]
```diff
@@ -1,217 +1,217 @@
...
+                    snakesNewLocation = gameSnake1.snakeMove(snakeDirection, false);
+                    mazeArray[snakesNewLocation.First().returnX(), snakesNewLocation.First().returnY()] = (int)Elements.snakeHead;
+
+                    mazeArray[food.xLocation, food.yLocation] = Food;
+                    break;
+            }       
+            return mazeArray;
+        }
+
+        private Point getNewHead(Direction snakeDirection)
```
Comment (danaterr) [inline Snake/BusinessLayer/Engine.cs]: use pascal case for methods even private methods.

File: Snake/BusinessLayer/Engine.cs [hunk 7]
```diff
@@ -1,217 +1,217 @@
...
+        public bool validateNewFoodLocation(Food newFood)
+        {
+            int x = newFood.xLocation;
+            int y = newFood.yLocation;
+
+            if ((x >= mazeLength) || (y >= mazeWidth))
+            {
+                return false;
+            }
+            if (mazeArray[x, y] == (int)Elements.mazeBody)
```
Comment (danaterr) [inline Snake/BusinessLayer/Engine.cs]: seems like your maze array should not be an array of ints but instead an array of enum types. then you would not have to keep casting

File: Snake/BusinessLayer/maze.cs [hunk 8]
```diff
@@ -1,137 +1,199 @@
...
     {
-        private const int BORDER = 1;
-        private const int NO_BORDER = 0;
+        private const int border = 1;
+        private const int noBorder = 0;
+        private enum direction { horizon, vertical };
         private int width, height;
-        private int[,] Maze;
-        Random random_number;
+        private int[,] maze;
```
Comment (danaterr) [inline Snake/BusinessLayer/maze.cs]: consider if int is the correct data type for the maze. what if it was nullable Elements instead?

File: Snake/BusinessLayer/maze.cs [hunk 9]
```diff
@@ -1,137 +1,199 @@
...
         }
 
 
-
-        public void Generate_All_Obstacles()
+        public void GenerateGridMaze()
         {
-            int Vertical_Line_Head_X = random_number.Next(0,height);
-            int Vertical_Line_Head_Y = random_number.Next(width/2,width);
+            for (int i = 0; i < height - 2; i = i + 4)
```
Comment (danaterr) [inline Snake/BusinessLayer/maze.cs]: consider if you should explain where the magic number 4 comes from. i can not tell by reading the code. an easy way to get around this is create a constant or else add a comment

**Comment by KuangyiXing** [comment]
Code Engine class is previous existing. Refactoring engine class is part of different stories. 
I made the maze.cs and maze_test.cs.

File: Snake/BusinessLayer/maze.cs (hunk 8 above)
Comment (KuangyiXing) [inline Snake/BusinessLayer/maze.cs]: every element is initialised to 0 when creating the maze

Comment (danaterr) [inline Snake/BusinessLayer/maze.cs]: enum can be 0. in fact unless you direct it other wise the first item is 0.

Comment (KuangyiXing) [inline Snake/BusinessLayer/maze.cs]: We will refactor it in another user story

**Pull request merged**